module/validation/receipt_validator.go

```diff
@@ -160,64 +161,203 @@ func (v *receiptValidator) Validate(receipts []*flow.ExecutionReceipt) error {
 	}
 
 	for i, r := range receipts {
-		err := v.validate(r, previousResult)
+		prevResult, err := fetchResult(r.ExecutionResult.PreviousResultID)
+		if err != nil {
+			return err
+		}
+
+		err = v.validateReceipt(r.Meta(), r.ExecutionResult.BlockID)
 		if err != nil {
 			// It's very important that we fail the whole validation if one of the receipts is invalid.
 			// It allows us to make assumptions as stated in previous comment.
 			return fmt.Errorf("could not validate receipt %v at index %d: %w", r.ID(), i, err)
 		}
+		err = v.validateResult(&r.ExecutionResult, prevResult)
+		if err != nil {
+			return fmt.Errorf("could not validate result %v at index %d: %w", r.ExecutionResult.ID(), i, err)
+		}
 	}
 	return nil
 }
 
-func (v *receiptValidator) validate(receipt *flow.ExecutionReceipt, getPreviousResult GetPreviousResult) error {
-	identity, err := identityForNode(v.state, receipt.ExecutionResult.BlockID, receipt.ExecutorID)
+func (v *receiptValidator) ValidatePayload(candidate *flow.Block) error {
+	// lookup cache to avoid linear search when checking for previous result that is
+	// part of payload
+	payloadExecutionResults := candidate.Payload.ResultsById()
+	// Build a functor that performs lookup first in receipts that were passed as payload and only then in
+	// local storage. This is needed to handle a case when same block payload contains receipts that
+	// reference each other.
+	// ATTENTION: Here we assume that ER is valid, this lookup can return a result which is actually invalid.
+	// Eventually invalid result will be detected and fail the whole validation.
+	fetchResult := func(previousResultID flow.Identifier) (*flow.ExecutionResult, error) {
+		prevResult, found := payloadExecutionResults[previousResultID]
+		if found {
+			return prevResult, nil
+		}
+
+		return v.fetchResult(previousResultID)
+	}
+
+	header := candidate.Header
+	payload := candidate.Payload
+
+	// Get the latest sealed block on this fork, ie the highest block for which
+	// there is a seal in this fork. This block is not necessarily finalized.
+	last, err := v.seals.ByBlockID(header.ParentID)
 	if err != nil {
-		return fmt.Errorf(
-			"failed to get executor identity %v at block %v: %w",
-			receipt.ExecutorID,
-			receipt.ExecutionResult.BlockID,
-			err)
+		return fmt.Errorf("could not retrieve parent seal (%x): %w", header.ParentID, err)
 	}
+	sealed, err := v.headers.ByBlockID(last.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve sealed block (%x): %w", last.BlockID, err)
+	}
+	sealedHeight := sealed.Height
 
-	err = ensureStakedNodeWithRole(identity, flow.RoleExecution)
+	// forkBlocks is used to keep the IDs of the blocks we iterate through. We
+	// use it to identify receipts that are for blocks not in the fork.
+	forkBlocks := make(map[flow.Identifier]*flow.Header)
+
+	// Create a lookup table of all the receipts that are already included in
+	// blocks on the fork.
+	forkLookup := make(map[flow.Identifier]struct{})
```

Comment:
maybe something to improve later, but I think technically, we only need to iterate through unfinalized block, because the lookup for receipts in finalized and unsealed blocks will never change.